tests/ops/util/test_ac_string.py

```diff
@@ -0,0 +1,259 @@
+    assert ac_str == ""
+
+    # Bad AC format
+    with pytest.raises(DecodeFidesStringError):
+        split_fides_string(
+            "CPz1hddPz1hddDxAAAENCZCgADgAAAAAAAAAAEBcABioAAA.YAAAAAAAAAA,100.1000"
+        )
+
+
+class TestBuildACString:
+    def test_opt_in_but_no_vendors(self):
+        contents = TCFExperienceContents()
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_in
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~"
+
+    def test_opt_out_preference(self):
+        contents = TCFExperienceContents(
+            tcf_vendor_consents=[
+                TCFVendorConsentRecord(id="gacp.1313"),
+                TCFVendorConsentRecord(id="gacp.231"),
+                TCFVendorConsentRecord(id="gacp.8844"),
+            ]
+        )
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_out
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~"
+
+    def test_opt_in_only_gvl_vendors(self):
+        contents = TCFExperienceContents(
+            tcf_vendor_consents=[
+                TCFVendorConsentRecord(id="gvl.1313"),
+                TCFVendorConsentRecord(id="gvl.231"),
+                TCFVendorConsentRecord(id="gvl.8844"),
+            ]
+        )
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_in
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~"
+
+    def test_opt_in_one_gacp_id(self):
+        contents = TCFExperienceContents(
+            tcf_vendor_consents=[
+                TCFVendorConsentRecord(id="gacp.12"),
+                TCFVendorConsentRecord(id="gvl.29"),
+            ]
+        )
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_in
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~12"
+
+    def test_badly_formatted_vendor_ids_skipped(self):
+        contents = TCFExperienceContents(
+            tcf_vendor_consents=[
+                TCFVendorConsentRecord(id="gacp.1313aasd"),
+                TCFVendorConsentRecord(id="gvl.asdf"),
+            ]
+        )
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_in
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~"
+
+    def test_build_opt_in_ac_string(self):
+        contents = TCFExperienceContents(
+            tcf_vendor_consents=[
+                TCFVendorConsentRecord(id="gacp.1313"),
+                TCFVendorConsentRecord(id="gacp.231"),
+                TCFVendorConsentRecord(id="gacp.8844"),
+            ]
+        )
+        ac_vendor_consents = build_ac_vendor_consents(
+            contents, UserConsentPreference.opt_in
+        )
+        assert build_ac_string(ac_vendor_consents) == "1~231.1313.8844"
+
+
+class TestDecodeACStringToPreferences:
```

Comment:
also nice tests here - could maybe use a case with multiple AC vendors specified in the AC string, just to get coverage on the multi-value parsing? (i.e. that the `.` delimiter is well handled). i know that's already implicitly covered above, but couldn't hurt to get a bit more coverage here specifically on this `decode_ac_string_to_preferences` function!

Reply:
good idea -